### ISSUE TITLE
[Dropdown] `set selected` trigger multiple times `onChange` callback when selecting multiple values

### ISSUE THREAD

**Issue by prudho** [fomantic/Fomantic-UI#165]
# Bug Report

## Steps to reproduce
1. Create a multiple selection dropdown
2. Use the `set selected` behavior to add at least two selected values
3. !?

## Expected result
`onChange` is fired once

## Actual result
`onChange` is fired multiple times

## Testcase
https://jsfiddle.net/kqbfpwm6/1/

**Comment by lubber-de** [comment]
:thinking: This seems to be wanted behavior instead of a bug:
It is triggered as much often as you have defined selected in your array. The same behaves for `onAdd` and `onRemove`.

You could argue, that `onChange` should only be triggered once per call, regardless how many entries you have chosen at once, while for `onAdd` it is more logical to trigger this for each entry.

I looked at the code and basically we can catch the call to `onChange` to not trigger every time, but only after the `.each` call to `.add.values`.

But currently for each callback there are 4 parameters defined: 
1. The dropdown.element
2. The overall current value of all selected entries
3. The current value which was added
4. The current text value which was added

When the `onChange` should only be called once after the whole iterations of each element, which values should be given to the callback Function then?

1. This is clear, the same as above
2. This is also clear, the same as above
3. This is unclear, should really all of the above values be given to the callback at once as a string?
4. Same as 3. but for the texts

I strongly believe this will break existing applications which might have implement it like in your fiddle, because their callback functions do not expect parameters 3. and 4. to have multiple values within a string.

**Comment by lubber-de** [comment]
According to the desired additional option for changing this behaviour only when needed, we at least have to decide what exactly should be delivered to the onChange callback as parameters 3 and 4. Then, i can easily create a PR for this

**Comment by lubber-de** [comment]
There are 2 options IMHO:
Suppose you had used `['dutch','english','french']` like in your fiddle.

Currently there are 3 calls to `onChange` with Parameters 3 and 4 as:
`'dutch','Dutch'`
`'english','English'`
`'french','French'`

With a PR there will be only 1 call (if wanted as a seprate option to  `dropdown()` with 2 possibilities IMHO:

a) Parameters 3 and 4 are given as strings
```html
'dutch english french','Dutch English French'
```
b) Parameters 3 and 4 are given as array objects
```html
['dutch','english','french'],['Dutch','English','French']
```

I vote for option b)

**Comment by ColinFrick** [comment]
@lubber-de Thanks for the investigation. @prudho is already working on this.

**Comment by prudho** [comment]
@lubber-de I indeed started looking at it. If you want to take a look at my WIP code, fell free to checkout it here: https://github.com/prudho/Fomantic-UI (as of today, the 2 last commits on my master branch since I am dumb...).

**Comment by lubber-de** [comment]
I did, good start so far :smiley: 

Comments:
- The existing "silent" option was/is already used to handle console outputs for debugging. This gets confusing if you also use this to handle the onChange behaviour. I would and add an additional parameter like "changeTriggerAlwaysOnMultiple: true/false" (ok, something shorter will also make it...)
- You have changed parameter types in case there where multiple entries given. It returns an array then instead of a string. This is basically fine and wanted, but there is a (from the other SUI issues wanted) dropdown-option missing which decides to use the old logic. Existing customer code will break now, because their onChange method might not be ready to retrieve an array yet. By default the code should behavee as before IMHO (and as suggested in all the SUI issues regarding this)
- You changed the amount of parameters for onAdd. The text-parameter is omitted for onAdd. Also a breaking change for existing code which expects that and now gets the item object instead
- Same for onChange. It now only retrieves 2 parameters. This makes totally sense, but will break existing code, so again, an additional parameter to dopdown is needed, so everybody only needs to change their code if they definately want to use/enable the new behaviour
- You need to cover the "remove selected" call also (this is still behaving as before in your current code)

**Comment by KennyZJ** [comment]
> @prudho I did, good start so far 😃
> 
> Comments:
> 
> * The existing "silent" option was/is already used to handle console outputs for debugging. This gets confusing if you also use this name as a parameter in the code to handle the onChange behaviour. I would use and  add an additional parameter like "changeTriggerAlwaysOnMultiple: true/false" (ok, something shorter will also make it...)
> * You have changed parameter types in case there were multiple entries given. It returns an array then instead of a string. This is basically fine and wanted, but there is a (from the other SUI issues wanted) dropdown-option missing which decides to use the old logic. Existing customer code will break now, because their onChange method might not be ready to retrieve an array yet. By default the code should behave as before IMHO (and as suggested in all the SUI issues regarding this)
> * You changed the ~amount of~ parameters for onAdd. The text-parameter is ~omitted~ always an empty string for onAdd. Also a breaking change for existing code which _does not_ expect~s~ that ~and now gets the item object instead~
> * Same for onChange, even on single selection. ~It now only retrieves 2 parameters.~ This makes totally sense, but will break existing code, so again, an additional parameter to dopdown is needed, so everybody only needs to change their code if they definately want to use/enable the new behaviour
> * You need to cover the "remove selected" call also (this is still behaving as before in your current code)

Agree, 'sliently' make dropdown selection without triggering 'onChange' event totally makes sense, especially when work with some other two-way bind frontend framework like angular.

**Comment by vincenccee** [comment]
hey guys, any update on this?

**Comment by prudho** [comment]
Sadly no 🙁

Managing to fix this bug without breaking anything is far more complicated than I previously though...

**Comment by cupcakedream** [comment]
Would love to see this change made +1

**Comment by truesteps** [comment]
If anyone still has an issue with this, you can use lodashes `debounce` or any debouncer to call the function you desire only on the last iteration :)

**Comment by ardamose123** [comment]
I'd like to tackle this. However, what I read in the discussion and how `onChange` works currently look different.

Currently, after using `set selected`, the `onChange` callback is fired once for each provided value that was not previously selected. The callback receives 3 parameters, `value`, `text` and `$choice`, as per the documentation in https://fomantic-ui.com/modules/dropdown.html#/settings.

The fix would be to have a configuration setting for `.dropdown({ ... })` to trigger `onChange` either once per affected option or just once per "batch operation" (`set selected`, `remove selected`). This setting will affect the data sent through the second parameter to the `onChange(value, text, $choice)` callback.

Am I on the right track here?

Another possible solution would be to have a completely new callback function (`onUpdate` or similar name) that provides the callback with the values and texts/labels as arrays and fires only once per batch operation. This shouldn't break any existing code in the end; people who want to have this new behavior would use the new callback instead of the existing `onChange`.
